Use native `Math` functions in the custom `log2` function

timvandermeij opened the pull request:
It is quite confusing that the custom function is called `log2` while it actually returns the ceiling value and handles zero and negative values differently than the native function.

To resolve this, we add a comment that explains these differences and make the function use the native `Math` functions internally instead of using our own custom logic. To verify that the function does what we expect, we add unit tests.

All browsers except for IE support `Math.log2` for quite a long time already (see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/log2). For IE, we use the core-js polyfill.

According to the microbenchmark at https://jsperf.com/log2-pdfjs/1, using the native functions should also be faster, in my testing almost six times as fast.

timvandermeij commented:
/botio-linux preview

timvandermeij commented:
/botio test

timvandermeij commented:
/botio-linux preview

Snuffleupagus commented:
Unfortunately `Math.ceil(Math.log2(x))` still isn't entirely compatible with the old helper function, compare e.g. the result of `x = 0` for the old/new code :-(

timvandermeij committed:
Use native `Math` functions in the custom `log2` function

It is quite confusing that the custom function is called `log2` while it
actually returns the ceiling value and handles zero and negative values
differently than the native function.

To resolve this, we add a comment that explains these differences and
make the function use the native `Math` functions internally instead of
using our own custom logic. To verify that the function does what we
expect, we add unit tests.

All browsers except for IE support `Math.log2` for quite a long time
already (see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/log2).
For IE, we use the core-js polyfill.

According to the microbenchmark at https://jsperf.com/log2-pdfjs/1,
using the native functions should also be faster, in my testing almost
six times as fast.

timvandermeij commented:
I have updated the patch to make sure that there are no difference in the implementations anymore.

timvandermeij commented:
/botio-linux preview

timvandermeij commented:
/botio test

Snuffleupagus commented on src/shared/util.js:
```diff
@@ -600,13 +600,14 @@ function string32(value) {
 
+// Calculate the base 2 logarithm of the number `x`. This differs from
+// the native function in the sense that it returns the ceiling value
+// and that `Infinity`/`NaN` values are treated as 0.
```
This line is really confusing me, since as far as I can tell `log2(Infinity) == Infinity` and `log2(NaN) == NaN` rather than `0`. What's correct here, the code or the comment?

timvandermeij replied:
Yes, now that I read it again I can see how it's a bit confusing. I rewrote it in the new commit, so hopefully it's better like that.

Snuffleupagus replied:
That's a lot clearer, thanks!

timvandermeij commented:
*I have updated the commit to adjust the comment above `log2`. No functional changes have been made.*

Snuffleupagus commented on test/unit/util_spec.js:
```diff
@@ -136,6 +136,19 @@ describe('util', function() {
+    it('handles values larger than zero', function() {
+      expect(log2(1)).toEqual(0);
+      expect(log2(2)).toEqual(1);
+      expect(log2(3)).toEqual(2);
```
Since our `log2` function doesn't behave like you would expect, based on the name, how about adding a test-case where the input is a floating point number?

timvandermeij replied:
Good suggestion! Done in the new commit.

Snuffleupagus reviewed (approved):
Seems fine to me, with one minor comment; thank you for the patch (and for adding more unit-tests)!

timvandermeij commented:
/botio unittest
